### ISSUE TITLE
Restructuring of folders/links

### ISSUE THREAD

**Issue by rarkins** [containerbase/base#213]
Today we see paths like this:

```
/usr/local/helm/3.7.1
/usr/local/dotnet
/home/ubuntu/.gem-global/bin
/home/ubuntu/.gem/ruby/3.0.0/bin
/usr/local/ruby/3.0.3/bin
/home/ubuntu/.cargo/bin
/usr/local/rust/1.57.0/bin
/usr/local/go/1.17.4/bin
/go/bin
/usr/local/elixir/1.13.0/bin
/home/ubuntu/.npm-global/bin
/usr/local/yarn/1.22.17/bin
/home/ubuntu/.local/bin
/usr/local/python/3.10.0/bin
/usr/local/poetry/1.1.12/bin
/usr/local/pnpm/6.23.1/bin
/usr/local/php/7.4.26/bin
/usr/local/node/14.18.2/bin
/usr/local/java/11.0.13+8/bin
/usr/local/gradle/6.9.1/bin
/usr/local/composer/2.1.14/bin
/home/ubuntu/bin
/home/ubuntu/bin
/usr/local/sbin
/usr/local/bin
/usr/sbin
/usr/bin
/sbin
/bin
```

Can we restructure how we install so that it can be like this?

```
/home/ubuntu/bin
/usr/local/sbin
/usr/local/bin
/usr/sbin
/usr/bin
/sbin
/bin
```

By this I'm hoping what can happen is that every time `install-tool` installs a version of a too, it symlinks it to `/usr/local/bin` or `/home/ubuntu/bin`.

e.g. `install-tool npm 8.0.0` might install it into `/home/ubuntu/.npm-global/bin` but then there's be a symlink from `/home/ubuntu/.npm-global/bin/npm` to `/home/ubuntu/bin/npm`.

If we don't have to think about paths and dynamic ENV injection then wouldn't it be much easier to know what's installed?

**Comment by viceice** [comment]
Sounds good, but i prefer our shell wrapper and use some fixed files to read appropriate env

**Comment by rarkins** [comment]
Are there advantages to having ~20 dirs in the $PATH though? I'm not sure which bits of what I want are conflicting with what you wrote

**Comment by viceice** [comment]
i now go the symlink path, should work for root and non-root for node tools.

**Comment by rarkins** [comment]
Do you want a separate issue for other tools, so we have something to track?

I think any that we allow runtime install for should be symlinked into the shared `bin` folders, because otherwise our path will keep growing and growing.

Here's latest `renovate/renovate` for reference:

```
/usr/local/dotnet
/home/ubuntu/.gem-global/bin
/home/ubuntu/.gem/ruby/3.0.0/bin
/usr/local/ruby/3.0.3/bin
/home/ubuntu/.cargo/bin
/usr/local/rust/1.57.0/bin
/usr/local/go/1.17.4/bin
/go/bin
/usr/local/elixir/1.13.0/bin
/home/ubuntu/.npm-global/bin
/home/ubuntu/.local/bin
/usr/local/python/3.10.0/bin
/usr/local/gradle/6.9.1/bin
/home/ubuntu/bin
/usr/local/dotnet
/home/ubuntu/.gem-global/bin
/home/ubuntu/.gem/ruby/3.0.0/bin
/usr/local/ruby/3.0.3/bin
/home/ubuntu/.cargo/bin
/usr/local/rust/1.57.0/bin
/usr/local/go/1.17.4/bin
/go/bin
/usr/local/elixir/1.13.0/bin
/home/ubuntu/.npm-global/bin
/home/ubuntu/.local/bin
/usr/local/python/3.10.0/bin
/usr/local/gradle/6.9.1/bin
/home/ubuntu/bin
/home/ubuntu/bin
/usr/local/sbin
/usr/local/bin
/usr/sbin
/usr/bin
/sbin
/bin
```

**Comment by rarkins** [comment]
Would it be cleaner to put everything in `/usr/local/buildpack/` and `/home/ubuntu/buildpack/`? Then add `/usr/local/buildpack/bin` and `/home/ubuntu/buildpack/bin` to `PATH`

**Comment by rarkins** [comment]
`jb` and `composer` are the other two tools we support for runtime install today

**Comment by viceice** [comment]
mostly done https://github.com/containerbase/buildpack/pull/216

**Comment by viceice** [comment]
> Would it be cleaner to put everything in `/usr/local/buildpack/` and `/home/ubuntu/buildpack/`? Then add `/usr/local/buildpack/bin` and `/home/ubuntu/buildpack/bin` to `PATH`

Sure, we can do this. will be a breaking, as we need to update `ENV PATH=...` in `Dockerfile`

**Comment by rarkins** [comment]
No breaking change if we stick to /usr/local/bin and /home/user/bin?

**Comment by viceice** [comment]
yes

**Comment by rarkins** [comment]
Perhaps not worth it then. Can we nest actual installs inside a buildpack folder though?

**Comment by viceice** [comment]
yes, thats possible.

**Comment by viceice** [comment]
Can do the refactoring after #218

**Comment by viceice** [comment]
current path on full image:

```sh
> echo $PATH | tr ":" "\n"
/usr/local/dotnet
/home/ubuntu/.gem-global/bin
/home/ubuntu/.gem/ruby/3.0.0/bin
/usr/local/ruby/3.0.3/bin
/home/ubuntu/.cargo/bin
/usr/local/rust/1.57.0/bin
/usr/local/go/1.17.4/bin
/go/bin
/usr/local/elixir/1.13.0/bin
/home/ubuntu/.local/bin
/usr/local/python/3.10.1/bin
/home/ubuntu/.npm-global/bin
/usr/local/gradle/6.9.1/bin
/home/ubuntu/bin
/usr/local/dotnet
/home/ubuntu/.gem-global/bin
/home/ubuntu/.gem/ruby/3.0.0/bin
/usr/local/ruby/3.0.3/bin
/home/ubuntu/.cargo/bin
/usr/local/rust/1.57.0/bin
/usr/local/go/1.17.4/bin
/go/bin
/usr/local/elixir/1.13.0/bin
/home/ubuntu/.local/bin
/usr/local/python/3.10.1/bin
/home/ubuntu/.npm-global/bin
/usr/local/gradle/6.9.1/bin
/home/ubuntu/bin
/home/ubuntu/bin
/usr/local/sbin
/usr/local/bin
/usr/sbin
/usr/bin
/sbin
/bin
```

**Comment by viceice** [comment]
We need to move tools below language:
* `/user/local/ruby/3.1.0/bundler/2.3.5`
* `/user/local/node/16.0.1/npm/8.3.1`

This will allow installing the same version of `npm` or `bundler` for different language versions without need to reinstall

**Comment by rarkins** [comment]
ok, as long as there's no file name conflicts because that directory name is already used

**Comment by viceice** [comment]
We should also make `/usr/local/<tool>` writable to root group, so we can add tools there as user, so it makes folder and path handling simpler.

Versions installed by root should be protected, so user can't modify them.

We also need to make `/usr/local/bin/<tool>` executable writable, so user can override that symlink / shell wrapper to a new version too.

We should write current tool version to `/usr/local/buildpack/versions/<tool>` and `/usr/local/buildpack/versions/<lang>.<tool>`, eg `/usr/local/buildpack/versions/node` and `/usr/local/buildpack/versions/node.npm`.

**Comment by Chumper** [comment]
> We need to move tools below language:
> 
>     * `/user/local/ruby/3.1.0/bundler/2.3.5`
> 
>     * `/user/local/node/16.0.1/npm/8.3.1`
> 
> 
> This will allow installing the same version of `npm` or `bundler` for different language versions without need to reinstall

Do you have an overview for me why that is needed?
I imagine that we can do this:
```
install-tool node X
install-tool npm Y
```
when we need another version of npm we do:
```
install-tool npm Z
```
which in the end will just install npm in the needed version or do I miss something?

If we want to do that, then we need to install both at the same time, like 
`install-tool node X npm Y`

or should we install `npm` generally into the node folder?
So that `install-tool npm Y` will be installed in the current npm folder `.../tools/node/X/npm/Y`?

**Comment by viceice** [comment]
The last option was my idea. Those where required if we process multiple repo's who require different language version.

Eg first repo needs node v16 and npm V8.
Second repo needs node v14 and npm V8.

Currently we would only install new node version and keep npm. Npm could have resolved peer deps to versions which don't support node v14, so later call can fail.

Not sure if this really happens in npm, but there are other languages and tools where it will happen for sure.

**Comment by Chumper** [comment]
Ok, so we need to make this configurable. 
Will see if I can get an example into the branch

**Comment by Chumper** [comment]
@viceice What tools that we currently have should be installed in a subfolder?

Currently I can think of:
* npm in node
* yarn in node
* pnpm in node
* bundler in ruby
* poetry in python
* composer in php (?)
* maven in java
* gradle in java
* scala in java (?)
* sbt in java (?)

or is this not needed for certain parts?

**Comment by rarkins** [comment]
I'm not sure if there's a foolproof way to know when a tool is dependent on the underlying version of the language, or not. Perhaps we need the concept of defining the compatibility, e.g. major/minor/patch. Major compatibility means we'd e.g. want to yarn if the major version of node changed (e.g. 14 -> 16). Minor would mean _not_ reinstalling Poetry if the minor version stayed the same (e.g. 3.8.0 -> 3.8.1). And "patch" can maybe mean "always reinstall" if any language difference (not sure if we have any?).

**Comment by Chumper** [comment]
We can make this as complex as we want, but maybe we have a way to provide a simple solution right now.
@viceice said that there are certain languages and tools where this is definitively required.

My current approach enables each tool to determine on its own if and where it wants to be installed.
So maybe we can start with the absolute basic tools and languages without a concept first and see how we can refactor this when the need arises?

**Comment by rarkins** [comment]
Sure, my concern was just if we ended up with hugely bloated runtime images because tools kept being reinstalled every time there's a patch bump of the language

**Comment by Chumper** [comment]
Do you run the images for extended amounts of time?
I imagine that images life for a few hours max. In that time there should a reasonable isntalls of languages and tools only.

The only thing I can imagine is that you run Renovate for a bunch of repos and hey all require different node and npm versions.
In that case we could have n*m installs instead of n+m.

**Comment by rarkins** [comment]
Right now the app and WhiteSource Renovate On-Premises essentially run indefinitely, but that might need to change if we bloat the containers too much at runtime

**Comment by Chumper** [comment]
Good to know, in that case we might need to think about how to solve this.

**Comment by viceice** [comment]
So the `PATH` is now smaller again:
```sh
>docker run --rm -it renovate/renovate sh -c 'echo $PATH | tr ":" "\n"'                                                              ─╯
/home/ubuntu/.cargo/bin
/home/ubuntu/.local/bin
/go/bin
/home/ubuntu/bin
/opt/buildpack/tools/python/3.11.1/bin
/home/ubuntu/.npm-global/bin
/home/ubuntu/bin
/usr/local/sbin
/usr/local/bin
/usr/sbin
/usr/bin
/sbin
/bin
```

I'm not sure if we can fix python and node global.

I've now implemented a way for ruby to install bundler and cocoapods ruby minor version depended.

eg Bundler v1.17.2 on Ruby v2.6.4:  `/opt/buildpack/tools/bundler/1.17.2/2.6.0`

We probably should close this issue now and open new for the version dependend tools. We can maybe do it like bundler now.
